### PR TITLE
Sign messages.

### LINKED ISSUE #14: Sign all messages.

### PR DESCRIPTION
**Pull request by afck**
Closes #14.

### CODE REVIEW

**Comment by afck** [comment]
Not sure whether this is the right way to do it, or whether signing and verifying should happen inside the `hydrabadger` module?

**Comment by c0gent** [comment]
This makes sense. Let me think about it just a bit before merging.

File: src/lib.rs [hunk 1]
```diff
@@ -306,17 +306,28 @@ impl<T: Contribution> From<WireMessageKind<T>> for WireMessage<T> {
...
     _t: PhantomData<T>,
 }
 
 impl<T: Contribution> WireMessages<T> {
-    pub fn new(socket: TcpStream) -> WireMessages<T> {
+    pub fn new(socket: TcpStream, our_key: SecretKey) -> WireMessages<T> {
         WireMessages {
             framed: Framed::new(socket, LengthDelimitedCodec::new()),
+            our_key,
+            their_key: None,
```
Comment (c0gent) [inline src/lib.rs]: A `WireMessages` struct is shared and used by all communications to all nodes. Saving anything node-specific isn't appropriate, node specific info is stored in the `Peer` struct.

File: src/lib.rs [hunk 2]
```diff
@@ -339,10 +350,22 @@ impl<T: Contribution> Stream for WireMessages<T> {
...
+                if let WireMessageKind::HelloRequestChangeAdd(_, _, pk) = msg.kind {
+                    if self.their_key.is_some() {
+                        return Err(Error::DuplicateHello);
+                    }
+                    self.their_key = Some(pk);
+                }
+                if !self.their_key.as_ref()
+                        .map_or(true, |pk| pk.verify(&s_msg.sig, &s_msg.message)) {
+                    return Err(Error::InvalidSignature);
+                }
```
Comment (c0gent) [inline src/lib.rs]: This additional processing should probably take place within `Handler` somewhere (where node specific info is available).

**Comment by afck** [comment]
Closing in favor of #16.

**Comment by afck** [comment]
> A `WireMessages` struct is shared and used by all communications to all nodes.

But new, separate instance is [created for each peer](https://github.com/poanetwork/hydrabadger/blob/51d526c51c72dc35666a9819ed103d77c7a2de0c/src/hydrabadger/hydrabadger.rs#L317), isn't it?

**Comment by c0gent** [comment]
Is it a bad sign if I'm forgetting how my own code works? :P

**Pull request merged**